viscy/data/hcs_ram.py

```diff
@@ -0,0 +1,346 @@
+                            f"transform type {type(aug)}."
+                        )
+                    self.train_patches_per_stack = num_samples
+        else:
+            self.augmentations = []
+
+        _logger.debug(f"Training augmentations: {self.augmentations}")
+        return list(self.augmentations)
+
+    def _fit_transform(self) -> tuple[Compose, Compose]:
+        """(normalization -> maybe augmentation -> center crop)
+        Deterministic center crop as the last step of training and validation."""
+        # TODO: These have a fixed order for now... ()
+        final_crop = [
+            CenterSpatialCropd(
+                keys=self.source_channel + self.target_channel,
+                roi_size=(
+                    self.z_window_size,
+                    self.yx_patch_size[0],
+                    self.yx_patch_size[1],
+                ),
+            )
+        ]
+        train_transform = Compose(
+            self.normalizations + self._train_transform() + final_crop
+        )
+        val_transform = Compose(self.normalizations + final_crop)
+        return train_transform, val_transform
+
+    def _set_fit_global_state(self, num_positions: int) -> torch.Tensor:
+        # disable metadata tracking in MONAI for performance
+        set_track_meta(False)
+        # shuffle positions, randomness is handled globally
+        return torch.randperm(num_positions)
+
+    def _setup_fit(self, dataset_settings: dict) -> None:
+        """
+        Setup the train and validation datasets.
+        """
+        train_transform, val_transform = self._fit_transform()
+        dataset_settings["channels"]["target"] = self.target_channel
+        # Load the plate
+        plate = open_ome_zarr(self.data_path)
+        # shuffle positions, randomness is handled globally
+        positions = [pos for _, pos in plate.positions()]
+        shuffled_indices = self._set_fit_global_state(len(positions))
+        positions = list(positions[i] for i in shuffled_indices)
+        num_train_fovs = int(len(positions) * self.split_ratio)
+
+        shared_dict = Manager().dict()
+        self.train_dataset = CachedDataset(
+            shared_dict,
+            positions[:num_train_fovs],
+            transform=train_transform,
+            **dataset_settings,
+        )
+        self.val_dataset = CachedDataset(
+            shared_dict,
+            positions[num_train_fovs:],
+            transform=val_transform,
+            **dataset_settings,
+        )
+
+    def train_dataloader(self) -> DataLoader:
+        sampler = ShardedDistributedSampler(self.train_dataset, shuffle=True)
+        return DataLoader(
+            self.train_dataset,
+            batch_size=self.batch_size // self.train_patches_per_stack,
+            num_workers=self.num_workers,
+            persistent_workers=bool(self.num_workers),
+            pin_memory=True,
+            shuffle=False,
+            timeout=self.timeout,
```

Comment:
@edyoshikun why is this needed?

Reply:
At the beginning I had to add this timeout if it was taking long time to cache. I don't think we need this and in fact if it's =0 it works fine